### PR TITLE
[POAE7-622] Plasma Input/Output Stream implementation

### PR DESCRIPTION
**Pull request by yeyuqiang**
First PR for plasma Input/Output Stream implementation

### CODE REVIEW

**Comment by yeyuqiang** [comment]
@winningsix @XinDongIntel

**Comment by winningsix** [comment]
@jikunshang

**Comment by winningsix** [comment]
Please associate it with the JIRA ID.

File: core/src/main/java/org/apache/spark/io/pmem/PlasmaInputStream.java [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    }
+    return true;
+  }
+
+  @Override
+  public synchronized int read() {
+    if (!refill()) {
+      return -1;
+    }
+    return buffer.get() & 0xFF;
```
Comment (jikunshang) [inline core/src/main/java/org/apache/spark/io/pmem/PlasmaInputStream.java:83]: should return byte?

File: core/src/main/java/org/apache/spark/io/pmem/PlasmaOutputStream.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+   *
+   * @param parentObjectId
+   * @param storeSocketName
+   */
+  public PlasmaOutputStream(String parentObjectId, String storeSocketName) {
+    this(parentObjectId, storeSocketName, DEFAULT_BUFFER_SIZE);
+  }
+
+  @Override
+  public void write(int b) {
```
Comment (jikunshang) [inline core/src/main/java/org/apache/spark/io/pmem/PlasmaOutputStream.java:66]: This API wants to wirte a byte or something else?

File: core/src/main/java/org/apache/spark/io/pmem/PlasmaOutputStream.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+      throw new IndexOutOfBoundsException();
+    } else if (len == 0) {
+      return;
+    }
+    int bytesToWrite = len;
+    while (bytesToWrite > 0) {
+      int remainBytesInBuf = buffer.remaining();
+      if (remainBytesInBuf <= bytesToWrite) {
+        buffer.put(b, off, remainBytesInBuf);
+        bytesToWrite -= remainBytesInBuf;
```
Comment (jikunshang) [inline core/src/main/java/org/apache/spark/io/pmem/PlasmaOutputStream.java:89]: how to handle bytes left here?

File: core/src/test/java/org/apache/spark/io/pmem/PlasmaOutputInputStreamSuite.java [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+    byte[] bytesWrite = prepareByteBlockToWrite(2.7);
+    PlasmaOutputStream pos = new PlasmaOutputStream(blockId, plasmaStoreSocket);
+    pos.write(bytesWrite);
+    pos.close();
+
+    ByteBuffer bytesRead = ByteBuffer.allocate(bytesWrite.length);
+    PlasmaInputStream pis = new PlasmaInputStream(blockId, plasmaStoreSocket);
+    byte[] buffer = new byte[DEFAULT_BUFFER_SIZE];
+    int len;
+    while ((len = pis.read(buffer)) != -1) {
```
Comment (jikunshang) [inline core/src/test/java/org/apache/spark/io/pmem/PlasmaOutputInputStreamSuite.java:110]: Do we have this API?  
 `public int read(byte[] b, int off, int len)`  off and len don't have a default value.

File: core/src/main/java/org/apache/spark/io/pmem/PlasmaInputStream.java (hunk 1 above)
Comment (yeyuqiang) [reply core/src/main/java/org/apache/spark/io/pmem/PlasmaInputStream.java:83]: & 0xFF will convert it to unsigned integer

File: core/src/main/java/org/apache/spark/io/pmem/PlasmaOutputStream.java (hunk 2 above)
Comment (yeyuqiang) [reply core/src/main/java/org/apache/spark/io/pmem/PlasmaOutputStream.java:66]: That's tha standard API from input stream

File: core/src/main/java/org/apache/spark/io/pmem/PlasmaOutputStream.java (hunk 3 above)
Comment (yeyuqiang) [reply core/src/main/java/org/apache/spark/io/pmem/PlasmaOutputStream.java:89]: The whole if .. else .. already handle the left bytes

File: core/src/test/java/org/apache/spark/io/pmem/PlasmaOutputInputStreamSuite.java (hunk 4 above)
Comment (yeyuqiang) [reply core/src/test/java/org/apache/spark/io/pmem/PlasmaOutputInputStreamSuite.java:110]: sure...These api all comes from outputstream

Comment (jikunshang) [reply core/src/test/java/org/apache/spark/io/pmem/PlasmaOutputInputStreamSuite.java:110]: Sorry I forgot this class extends InputStream.

File: core/src/main/java/org/apache/spark/io/pmem/PlasmaOutputStream.java (hunk 2 above)
Comment (yeyuqiang) [reply core/src/main/java/org/apache/spark/io/pmem/PlasmaOutputStream.java:66]: Currently I dont think we will use this method to write, but just leave it here.

**Comment by jikunshang** [comment]
LGTM.

**Pull request merged**